Instructor: edit rubric question: allow moving columns #3910

VamsiSangam opened the pull request:
Fixes #3910 

**Demo**

![4lyw8jr6rf](https://cloud.githubusercontent.com/assets/11662001/26622541/625142dc-4608-11e7-8f74-8227ea8ba514.gif)

**Outline of solution**

- Moved buttons to the bottom by creating a new `<tr>` before the end of table body.
- Added `feedbackQuestionRubricEditFormTableOptions.html` and `feedbackQuestionRubricEditFormTableOptionsFragment.html` in `src/main/resources` which has HTML code related to the buttons at the bottom.
- Added code in `teammates.common.datatransfer.questions.FeedbackRubricQuestionDetails` and `teammates.common.util.Templates` which supports the above template files.
- Addded `swapRubricCol` in `questionRubric.es6`. It takes 3 parameters, column index, question number and a boolean `isSwapLeft` which indicates if this column is supposed to be swapped left or right.
- The left arrow button has `onclick="swapRubricCol(${col}, ${questionIndex}, true)"` and vice versa.
- `swapRubricCol` first checks validity of input, then computes the index of the column to swap, which is index of column to its left/right. This is done by using `data-col` attribute of `td` tags in the bottom row of buttons. Then swaps column contents by swapping choices and weights manually, and then swapping each row through a loop. The value in `textarea` tags is swapped.
- Clicking "move left" for the leftmost column and vice versa doesn't do anything for now. It is simply ignored.
- Added tooltips for buttons accordingly.

Can refer to #7408 for some additional info. Need help with variable names :sweat_smile: . If the current implementation is acceptable, I will move to write test code for this.

chowyb commented:
@VamsiSangam could you provide screenshot(s) of the tooltips for the buttons? It doesn't seem to appear in the GIF.

VamsiSangam commented:
Sorry, my bad. I thought I added the tooltips. Will update the code :sweat_smile:

damithc commented:
I'm curious as to why (in the gif) the text gets entered into different places of the text box, interestingly, with increased level of indentation. Looks like the text box already has some white spaces in it.

VamsiSangam commented:
@damithc sir, the space is due to #7382

VamsiSangam commented:
Current state of UI -

![ckt3rcvrza](https://cloud.githubusercontent.com/assets/11662001/26630321/d233ccd0-4623-11e7-86a8-8972ebb3bc8d.gif)

VamsiSangam commented:
Ready for review.

chowyb commented:
Tried running your code locally and things seem fine, but a few suggestions:

1. Tooltip text could say "Move column left/right" (ie. remove the "to")
2. If possible, the move left button for the leftmost column and the move right button for the rightmost column could be disabled. This would be better from a UI perspective.
3. (Lower priority) The buttons at the bottom may span multiple rows if there are a lot of columns (&ge;7 columns for my current resolution of 1366 x 768). You might want to see if there is a way to avoid this; however if this cannot be done without shrinking the buttons, then this can be left as is since the ability to easily click the buttons is more important.

Otherwise, the feature seems fine. I haven't reviewed for code quality yet however, as the code may change over the course of addressing these suggestions.

VamsiSangam commented:
Noted @chowyb :+1: . Will make the changes.

VamsiSangam commented:
UI currently -

![nl20d7ssvy](https://cloud.githubusercontent.com/assets/11662001/26682752/32244e54-46ff-11e7-87c4-b4f23c3a4c3a.gif)

The red status messages in the start are unrelated. I was trying something out.

VamsiSangam commented:
Ready for review.

chowyb commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java:
```diff
@@ -425,6 +425,24 @@ public String getQuestionSpecificEditFormHtml(int questionNumber) {
+            String tableBodyCell = Templates.populateTemplate(tableOptionsFragmentTemplate,
+                    Slots.QUESTION_INDEX, questionNumberString,
+                    Slots.COL, Integer.toString(i),
+                    Slots.ROW, Integer.toString(11)); // Considering the options row to be row number 11. Removable?
```
Not very fond of the "magic number" appearing here. If necessary, create a new constant for it.

VamsiSangam replied:
This is a part of some experiments. It is not used now as there is no `${row} in `feedbackQuestionRubricEditFormTableOptionsFragment.html`. Will remove it :+1:

chowyb commented on src/main/resources/feedbackQuestionRubricEditFormTableOptionsFragment.html:
```diff
@@ -0,0 +1,16 @@
+<td class="align-center rubricCol-${questionIndex}-${col}" data-col="${col}">
+    <div class="btn-group">
+        <button type="button" class="btn btn-default" id="rubricMoveChoiceLink-${questionIndex}-${col}-l" onclick="swapRubricCol(${col}, ${questionIndex}, true)"
+            data-toggle="tooltip" data-placement="top" title="Move column left">
```
Lines that are intended to be a continuation of a previous line should have 8 spaces, to avoid confusion between this and inner elements.

VamsiSangam replied:
I didn't know that :sweat_smile: . Will do. 

Side note, can't we check something like this in the style checks?

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -46,7 +46,7 @@ function addRubricRow(questionNum) {
 
     // Row to insert new row after
-    const lastRow = $(`#rubricEditTable-${questionNum} tr:last`);
+    const lastRow = $(`#rubricEditTable-${questionNum} tbody tr:nth-last-child(2)`);
```
This variable might need renaming as it is no longer the last row.

VamsiSangam replied:
Will do. :+1:

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -118,12 +109,44 @@ function addRubricCol(questionNum) {
+                </button>
+                <button type="button" class="btn btn-default" id="rubricMoveChoiceLink-${questionNum}-${newColNumber - 1}-r"
+                    onclick="swapRubricCol(${newColNumber - 1}, ${questionNum}, false)" data-toggle="tooltip"
+                    data-placement="top" title="Move column right">
```
Indentation of 8 spaces applies here too.

In addition, as I'm not too familiar with current plans for the JS side, tagging @thenaesh here: is there a better way of doing this (ie. not writing HTML in JS, and handling `onclick` and similar events)?

thenaesh replied:
Yes, it's preferred to not do this (though in this case it's not as bad since `swapRubricCol` is in the same file). Put the `onclick` in a separate function.

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -179,6 +203,91 @@ function removeRubricCol(index, questionNum) {
+function swapRubricCol(colIndex, questionNum, isSwapLeft) {
+    if ($(`#rubricEditTable-${questionNum}`).length === 0
+        && $(`.rubricCol-${questionNum}-${colIndex}`).length === 0
+        && typeof isSwapLeft !== 'boolean') {
```
8 spaces here too.

chowyb commented on src/main/resources/feedbackQuestionRubricEditFormTableOptionsFragment.html:
```diff
@@ -0,0 +1,16 @@
+            data-toggle="tooltip" data-placement="top" title="Move column left">
+            <span class="glyphicon glyphicon-arrow-left"></span>
+        </button>
+        <button type="button" class="btn btn-default" id="rubricRemoveChoiceLink-${questionIndex}-${col}" onclick="removeRubricCol(${col}, ${questionIndex})"
```
Missed this earlier: any `id` or `class` should also be in lowercase, with hyphens between words.

VamsiSangam replied:
But this style seems to have been used quite throughout the page :thinking:

chowyb replied:
@VamsiSangam yes, unfortunately older code does not seem to follow the style. An attempt was made to change the style of older code (see #5679 for an example) but was ultimately not done, and given the current level of activity from core members, GSoC students and additional contributors, it may be difficult to fix this right now without causing conflicts. Still, it would be better to follow the conventions now to reduce the amount of fixing that will need to be done later on.

VamsiSangam commented:
Oh hey @chowyb , I still need to write tests :stuck_out_tongue_closed_eyes: . Please add `s.Ongoing` :grin: . But I still want to know what @damithc sir has to say about the UI.

damithc commented:
No objections for the current code. @VamsiSangam please proceed to write tests.

chowyb commented:
@VamsiSangam whoops, my bad =P

VamsiSangam commented:
Outline of new changes -

- Removed inline JS.
- Renamed IDs to use the snake-naming-convention.
- Bug: new rubric question didn't have leftmost and rightmost buttons disabled. Fixed it and added related tests for it.

Ready for review.

chowyb commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java:
```diff
@@ -438,7 +455,8 @@ public String getQuestionSpecificEditFormHtml(int questionNumber) {
                 Slots.RUBRIC_TOOLTIPS_ASSIGN_WEIGHTS, Const.Tooltips.FEEDBACK_QUESTION_RUBRIC_ASSIGN_WEIGHTS,
-                Slots.RUBRIC_PARAM_ASSIGN_WEIGHTS, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHTS_ASSIGNED);
+                Slots.RUBRIC_PARAM_ASSIGN_WEIGHTS, Const.ParamsNames.FEEDBACK_QUESTION_RUBRIC_WEIGHTS_ASSIGNED,
+                Slots.RUBRIC_TABLE_OPTIONS, tableOptionsHtml.toString()); // resume here! fix compilation errors
```
This comment was probably not intended to be added =P

VamsiSangam replied:
Yeah I forgot to remove it! :stuck_out_tongue_closed_eyes:

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -118,12 +109,51 @@ function addRubricCol(questionNum) {
+                    <span class="glyphicon glyphicon-remove"></span>
+                </button>
+                <button type="button" class="btn btn-default" id="rubric-move-col-right-${questionNum}-${newColNumber - 1}"
+                        onclick="swapRubricCol(${newColNumber - 1}, ${questionNum}, false)" data-toggle="tooltip"
```
`onclick` inline JS still seem to be introduced here. If the concern is with new elements not having an associated event handler attached to them, consider using delegated events instead.

Note that this is mainly for the left and right buttons; the delete button is assumed to be handled by, or will be handled by, a PR for the `onclick` events.

VamsiSangam replied:
Sorry, I already added the `onclick` handlers. This is old code, forgot to remove. When any change is made in HTML, I need to update the `.html` file in resource as well as the code here. So I keep forgetting it :sweat_smile:

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -46,8 +46,8 @@ function addRubricRow(questionNum) {
     // Row to insert new row after
-    const lastRow = $(`#rubricEditTable-${questionNum} tr:last`);
-    $(newRubricRow).insertAfter(lastRow);
+    const secondLastRow = $(`#rubricEditTable-${questionNum} tbody tr:nth-last-child(2)`);
```
While we're at it, maybe we should prefix this variable name with a $ ie `$secondLastRow`. This was a convention that was made to distinguish jQuery objects from other variables, though older code may not have been updated to follows this.

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -118,12 +109,51 @@ function addRubricCol(questionNum) {
+            </div>
+        </td>`;
+
+    const lastTd = $(`#rubric-options-row-${questionNum} td:last`);
```
$ prefix applies here too.

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -118,12 +109,51 @@ function addRubricCol(questionNum) {
+    });
+
+    $(`#rubric-move-col-right-${questionNum}-${newColNumber - 1}`).tooltip({ container: 'body' });
+    $(`#rubric-move-col-right-${questionNum}-${newColNumber - 1}`).click(() => {
```
Consider extracting the left and right buttons as variables first. Remember to use the $ prefix if necessary.

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -179,6 +210,91 @@ function removeRubricCol(index, questionNum) {
+    } else {
+        // trying to swap right most column to right
+        return;
+    }
```
I was going to suggest rewriting this so it would look like

```javascript
if (isSwapLeft) {
    // left stuff
} else {
    // right stuff
}
```

but I realised it adds a redundant level of conditionals, so this should be fine.

VamsiSangam replied:
I remember writing it like that, something similar to this -

``` javascript
if (isSwapLeft) {
    if (condition) {
        // code
    } else {
        // trying to swap left most column to left
    }
} else {
    if (condition) {
        // code
    } else {
        // trying to swap right most column to right
    }
}
```

Then the style check failed saying `lonely if inside else` so I had to make it and `else if`.

chowyb commented on src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java:
```diff
@@ -325,6 +327,79 @@ public void testEditQuestionAction() throws Exception {
+        assertTrue(feedbackEditPage.moveRubricColLeft(1, 5));
+        assertTrue(feedbackEditPage.moveRubricColLeft(1, 3));
+        assertTrue(feedbackEditPage.moveRubricColLeft(1, 2));
+        assertTrue(feedbackEditPage.moveRubricColLeft(1, 1));
```
Without `assertTrue(feedbackEditPage.moveRubricColLeft(1, 4));` it might not be moving the last column to the first position.

VamsiSangam replied:
Actually I added a column (4) and deleted it in the code just above this. This tests if buttons of JS added columns are working properly.

chowyb commented on src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java:
```diff
@@ -325,6 +327,79 @@ public void testEditQuestionAction() throws Exception {
+        // move second column to last
+        assertTrue(feedbackEditPage.moveRubricColRight(1, 1));
+        assertTrue(feedbackEditPage.moveRubricColRight(1, 2));
+        assertTrue(feedbackEditPage.moveRubricColRight(1, 3));
```
I think the second column (index 1) seems to reach the fifth column (index 4) after this.

VamsiSangam replied:
Yes that is the intention. I wanted to swap column 0 and column 5 (column index 4 is deleted). So the code above these lines brings column 5 to 0 and these lines brings column 1 (which has column 0's content) to column 5.

Please note that column is not literally picked and placed left or right. The values inside the input/textarea tags are only swapped. So the column indices remain the same (0, 1, 2, 3, 5).

chowyb replied:
Ah alright, missed it, should be fine then 👍

VamsiSangam commented:
My `AllJsTests` is failing due to 23.99% coverage which less then the required minimum 24% -

![capture](https://user-images.githubusercontent.com/11662001/27001290-194b671e-4de4-11e7-96cd-a9868c1bb889.JPG)

Can anyone help me on how to solve this?

chowyb commented:
@wkurniawan07 might be able to help with this one, as I'm not very familiar with JS side testing.

wkurniawan07 commented:
@VamsiSangam you can reduce it to 23 for the time being. We haven't got the chance to _actually_ enforce JS coverage yet, for quite a long reason.

VamsiSangam commented:
Ready for review.

chowyb commented on src/main/webapp/js/instructorFeedbackEdit/questionRubric.es6:
```diff
@@ -176,6 +209,91 @@ function removeRubricCol(index, questionNum) {
+function swapRubricCol(questionNum, colIndex, isSwapLeft) {
+    if ($(`#rubricEditTable-${questionNum}`).length === 0
+            && $(`.rubricCol-${questionNum}-${colIndex}`).length === 0
+            && typeof isSwapLeft !== 'boolean') {
```
I probably missed this earlier, but looking at it again, should the `&&` be replaced by `||`? It appears that this particular conditional checks for failing conditions, in which case having any of them be true should be sufficient to do an early return.

VamsiSangam commented:
Ready for review.

VamsiSangam commented:
Ready for review.

thenaesh commented:
Remember to fix the new merge conflicts. :stuck_out_tongue: :bomb:

VamsiSangam commented:
Ready for review.

VamsiSangam commented:
@thenaesh reminder :)

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java:
```diff
@@ -425,6 +425,23 @@ public String getQuestionSpecificEditFormHtml(int questionNumber) {
+        String tableOptionsTemplate = FormTemplates.RUBRIC_EDIT_FORM_TABLE_OPTIONS;
+        String tableOptionsFragmentTemplate = FormTemplates.RUBRIC_EDIT_FORM_TABLE_OPTIONS_FRAGMENT;
+
+        for (int i = 0; i < numOfRubricChoices; ++i) {
```
`i++`? I don't think we use `++i` anywhere else in the main codebase.

VamsiSangam replied:
Nice catch. Sorry for that, old habits die hard :sweat_smile:

VamsiSangam commented:
Ready for review.

VamsiSangam commented:
Ready for review. @thenaesh reminder :)

wkurniawan07 commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -10,6 +10,95 @@ import {
 
+import {
+    extractQuestionNumFromEditFormId,
+} from '../main/instructorFeedbackEdit.es6';
```
Let's not do this. `common/*.es6` should be independent from `main/*.es6`.

VamsiSangam commented:
Ready for review.

VamsiSangam commented:
@damithc sir, addressing questions in your review,

>Tests feel too much 'copy-paste'. This is a common problem even in existing tests. See if you can write tests at a higher level of abstraction

I do agree with this and writing tests at a higher level of abstraction would allow us to write tests in lesser LOC and perhaps we could write more tests. But I feel it deviates a little from the scope of this issue, eg. if I implement `fillRubricColumn()`, it could probably be used it in another test case as well to reduce LOC. Also, I think targeting all the code in `FeedbackRubricQuestionUiTest.java` and trying to convert it to use greater abstraction methods seems to be a better approach than restricting ourselves to the moving columns code. So I feel this can be more effectively address in a separate issue. I'll gladly take it :grin: Please let me know your thoughts on this.

> How does moving columns affect existing responses?

This one blew my mind :sweat_smile: Why didn't this strike me before? :stuck_out_tongue_closed_eyes: Given the current implementation, the existing responses are not touched. So the results will look jeopardized if the instructor swaps `Agree` and `Disagree`. A simple choice would be to delete all the responses, but on the server side it is actually very very hard to tell if the instructor moved the columns. Move column logic is simply based on swapping the `textarea` values. So, in the server's point-of-view, it is simply an edit operation.

Move column buttons were proposed to be added to save the instructor effort of manually cut pasting a column values while swapping them. Now by adding "move" column buttons, we might unintentionally give faith to the instructor that the responses will also rearrange themselves. So I think there are 2 solutions to fix this issue -

1. Move column buttons will only be visible to a new question, not an existing question.
2. Move column buttons will only be visible if there are no existing responses.
3. Display a notice box saying something like "Moving columns does not change existing responses, use with caution".

Suggestions are most welcome :sweat_smile:

damithc commented:
1. It's fine if you do a more thorough testing in a PR immediately following this.
2. How about this? If a rubric question has responses, clicking on a swap button will show a message saying that swapping requires deletion of existing responses and offer buttons <kbd>Cancel</kbd><kbd>Delete Responses</kbd>

VamsiSangam commented:
1. I will create another issue for the improving test code and work on it after this PR is merged.
2. I have a few questions regarding the suggested approach. What would `Cancel` do? Revert the changes of move column button? And let's say the instructor pressed move column button, we show this message and offer options, what if the instructor decides to ignore `Delete Responses`, moves column and saves question anyway? What should we do about the responses then? If possible, could you please provide a mock UI for this? I mean do you expect a modal or a status message kind-of box or anything else?

damithc commented:
> I have a few questions regarding the suggested approach. What would Cancel do? Revert the changes of move column button? And let's say the instructor pressed move column button, we show this message and offer options, what if the instructor decides to ignore Delete Responses, moves column and saves question anyway? What should we do about the responses then? If possible, could you please provide a mock UI for this? I mean do you expect a modal or a status message kind-of box or anything else?

How's this? We show the dialog when the user clicks a swap button, _before_ actually swapping the columns. If user chooses `cancel`, no swapping will happen and the responses will not be deleted. If the user chooses `delete`, we delete the responses using an ajax request and then do the swap as requested.

VamsiSangam commented:
Understood sir, working on it. :smile:

VamsiSangam commented:
New UI -

![idox2iktnb](https://user-images.githubusercontent.com/11662001/27786855-776d386c-6000-11e7-86e8-51923608a3fd.gif)

Features -

- If the question has existing responses and user clicks move column button, modal shows up asking user to delete existing responses first.
- If user chooses `cancel`, nothing is done.
- If user chooses `Delete responses`, an AJAX request is sent to a new URL which deletes the responses and refreshes the respondent list.
- If successfully deleted, a confirmation modal shows up saying deletion was successful, otherwise a try-again message is displayed.
- After the user deleted the existing responses he/she can freely move columns.

Note -

1. This requested feature -
> If the user chooses delete, ... and **then do the swap as requested**.

could not be implemented. If the user chooses delete, the responses are deleted, and only after this his/her move column clicks will move the columns. The initial move column click (which triggers deletion) does not move columns. Reason - `swapRubricCol()` calls `checkExistingResponsesBeforeMovingCols()` which calls `deleteExistingResponses()`. If we were to implement this feature, `deleteExistingResponses()` would have to call `swapRubricCol()` on the initial parameters. The call can be made, but due to this cycle, I always got a style error eg. -

> `swapRubricCol` is used before it is defined in `deleteExistingResponses`

And if I put `swapRubricCol` ahead of `deleteExistingResponses`, I got and style error like -

> `checkExistingResponsesBeforeMovingCols` is used before it is defined in `swapRubricCol`

This style error couldn't be avoided due to the cycle method calls formed.

2. If the user changes number of rows/columns, then user can freely move columns without a "Delete existing responses" restriction, because changing number of rows/columns will delete the existing responses anyway, and the user is also warned of this before save.

Ready for review.

VamsiSangam commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -1,15 +1,162 @@
+}
+
+function swapRubricCol(questionNum, colIndex, isSwapLeft) {
+    checkExistingResponsesBeforeMovingCols(questionNum);
```
I was just following what @damithc sir suggested 

> We show the dialog when the user clicks a swap button

damithc replied:
> There's a "Save Changes" button that an instructor will need to click before the columns are actually swapped. The existing responses should only be deleted at that point.

We can do that too, as long as the warning is given at the beginning itself. However, I thought of deleting the responses at be beginning because in case something goes wrong with the delete operation (e.g. timeout), the instructor is going to lose all the edits he did to the question. Thoughts?

VamsiSangam replied:
If you really want what @thenaesh suggested, it is same as that of adding/removing rows/columns. When an instructor add/removed rows/columns, we show a `Warning: Existing responses will be deleted by your action` modal and ask for confirmation. If you want the same thing here, then it would have been just one line of code (set `editstatus` attribute to `mustDeleteResponses`) instead of my last 12 or so commits :rofl:

damithc replied:
I didn't realize the mechanism already exists. We certainly don't want two mechanisms in the same question either. So how about we move those 12 commits to some other branch and use the existing mechanism in this PR? After that, we can consider moving the entire question to a more user friendly mechanism where warning and possibly the deletion is done at the beginning rather than an the end.

VamsiSangam replied:
@thenaesh , just want to confirm. Changing `editstatus` attribute to `mustDeleteResponses` of the question's `form` tag will prompt the required `Warning: Existing responses will be deleted by your action` dialog. Shall I proceed with this approach?

If yes, I plan to run -

```
git rebase -i HEAD~15
```
and then remove the last 14 commits, reverting my branch to commit [fbf8f60](https://github.com/TEAMMATES/teammates/pull/7415/commits/fbf8f60b89c5dc92c6c66634d938a8bd07533dbe), and then force push. Is this the correct way undoing commits? (I'm still a git beginner :sweat_smile: ) . All work done in this branch so far is saved to another [branch](https://github.com/VamsiSangam/teammates/tree/3910-move-cols-deletion-before-moving) for safekeeping.

thenaesh replied:
@VamsiSangam That should be fine. You just need to `git reset --hard HEAD~14` to move 14 commits back.

VamsiSangam commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -1,15 +1,162 @@
+    }
+}
+
+function swapRubricCol(questionNum, colIndex, isSwapLeft) {
```
Taking the jQuery object of rubric edit table will only make the code more complicated. Since individual input elements have their unique IDs which can be accessed by the use of question number, the code will be much simpler if we have the question number.

VamsiSangam commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -1,15 +1,162 @@
+    }
+
+    // check the second last column right button, should be enabled
+    if ($optionColumns.length > 2) {
```
Actually `disableCornerMoveRubricColumnButtons()` doesn't work that way, it has nothing to do with `$optionColumns.length < 2`. There are just 3 conditions to check -

- Rightmost button should be disabled.
- Leftmost button should be disabled.
- Move right button of 2nd last column should be enabled (when instructor adds a new column).

Any further checks would only be redundant. So in my method I simply wrote the code to check for these 3 conditions only.

VamsiSangam commented:
Ready for review.

thenaesh commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -10,6 +10,110 @@ import {
+        const $secondlastColRightBtn = $secondlastCol.find(`#rubric-move-col-right-${questionNum}-${secondlastColIndex}`);
+
+        $secondlastColRightBtn.prop('disabled', false);
+    };
```
One question: why is the second last column's right button treated specially?

Apart from that, two things:
1. There's some repetition here, which can be refactored out.
1. The `check...` naming suggests some sort of test that returns a boolean value. Perhaps calling it `setCornerButtonState` would be better.

VamsiSangam replied:
> why is the second last column's right button treated specially?

For a special case -
  - If a new column is added, then 2nd last column's right button would be disabled, it must be enabled.

> There's some repetition here, which can be refactored out.

Where is the repetition?

>  Perhaps calling it setCornerButtonState would be better.

Renaming them to `disableMoveLeftOfFirstCol`, `disableMoveRightOfLastCol`, `enableMoveRightOfSecondLastCol`

thenaesh commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -10,6 +10,110 @@ import {
+    }
+}
+
+function moveRubricCol(questionNum, colIndex, isMoveLeft) {
```
`moveRubricColIfPossible`

thenaesh commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -10,6 +10,110 @@ import {
+    return $(`#rubricDesc-${questionNum}-${row}-${col}`);
+}
+
+function swapRubricCol(questionNum, colIndex, isSwapLeft) {
```
`function moveRubricCol(questionNum, colIndex, isMoveLeft)`

If you insist on not taking in two elements and swapping them, and instead would rather take in a single element that can be moved in the left or right direction, then what is being done here is not really swapping but moving.

VamsiSangam replied:
I refactored code a bit. You might find the new version more pleasing. :sweat_smile:

VamsiSangam commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -10,6 +10,110 @@ import {
+    // swap rubric choices
+    let $currentValue = getRubricChoiceElem(questionNum, colIndex);
+    let $swapValue = getRubricChoiceElem(questionNum, swapColIndex);
+    swapValues($currentValue, $swapValue);
```
This is pretty much what you suggested in your review comment just above - 
``` js
const swap = (elemSelector, questionNum, indexOfFirstCol, indexOfSecondCol) => {
    const $a = elemSelector(questionNum, indexOfFirstCol);
    const $b = elemSelector(questionNum, indexOfSecondCol);
    const $temp = $a.val();
    $a.val($b.val());
    $b.val($a.val());
};
```

I am going with this.

VamsiSangam commented on src/main/webapp/dev/js/common/questionRubric.es6:
```diff
@@ -10,6 +10,110 @@ import {
+        const temp = $a.val();
+        $a.val($b.val());
+        $b.val(temp);
+    };
```
It may not look good actually. Because, rubric desc needs `row` also. I will use optional parameters to make the code look good. Please review on the new code once I push it.

VamsiSangam commented:
@thenaesh refactored code. Please review.

thenaesh commented:
@chowyb

VamsiSangam commented:
Nice catch @chowyb :+1: 

Somewhere along the line, I must have altered the logic for computing question number. Using `data-qnnumber` attribute of the `form` element is the right way to do. In my buggy code, I was using value of `<input name="questionnumber">`, that value is not `-1` for a new question, which is where the bug comes in.

I did write tests for it, but they magically passed because I modify the HTML test files concerning them very frequently (I need to update them every time I sync my branch). So I guess in some the bug crept to the HTML test files.

Made the necessary changes.

Ready for review.

chowyb commented:
@VamsiSangam Just tried checking for the behaviour for deleting responses if a row is swapped, and would just like to check: is the deletion of responses only done for adding/removing columns? I am unable to trigger the warning to delete responses when doing swaps only.

VamsiSangam commented:
@chowyb you are right. Upon moving the columns, delete responses modal should pop up, as discussed with @damithc sir and @thenaesh . I completely forgot about that. Apologies for requesting review so many times. :sweat: Will add it.

VamsiSangam commented:
Ready for review.

VamsiSangam commented:
@chowyb ready for review.

VamsiSangam commented:
@damithc sir, your thoughts on the deletion?

damithc commented:
> Seems okay, although I have concerns that the deletion is now too strict (eg. editing a rubric choice without doing swapping, such as by adding one character to the choice) will cause deletion as well. However, it would probably be best for @damithc to decide on what the best behaviour would be.

No, simple edits should not delete responses. e.g. fixing a typo or adding more details into a cell. Only swapping/adding/deleting columns/rows should delete responses because those operations cause the existing data to be unusable.

VamsiSangam commented:
@chowyb and @damithc sir, the current `isChangesRequiresResponseDeletion` logic in the master branch does not allow simple edits to rubric choice (changes to rubric description is okay). So my previous logic -

``` java
if (!this.rubricChoices.equals(newRubricDetails.rubricChoices)) {
    return true;
}
```

doesn't actually make things "very strict", as the restrictions were already there to begin with. So I still think my `equals()` based logic is right.

Also please note that if you want to allow edits to rubric choices, but delete responses for swaps, there's no simple way to do it. Mainly because we can't differentiate between a swap and an edit operation in the server. Or at least if we wanted to, the code will get very complicated.

I think we should stick with `equals()` based approach. Please let me know about your thoughts on this.

(Ignore my latest commit for this comment's discussion).

damithc commented:
> Also please note that if you want to allow edits to rubric choices, but delete responses for swaps, there's no simple way to do it.

Swap/add/delete operations are done using UI buttons, right? Can't we keep track of whether a swap/add/delete was done? Anyway, we can do it as a follow up PR.

VamsiSangam commented:
> Can't we keep track of whether a swap/add/delete was done?

Sir this might seem like a good idea, but things get complicated if the instructor performs swaps, then edits or vice versa, i.e, a mixture of edits/swaps. Plus, we might end up sending too much data to the server if we were to send all the tracked UI actions and on top of that simulating user's actions in the server will be complicated. So, I don't think "tracking" is an elegant or a scalable solution.

 However, I will create an issue for that after this PR, discussion over this might give us a better solution.

VamsiSangam commented:
@chowyb your thoughts on this? Should I stick with the `equals` approach?

damithc commented:
It's fine to use the stricter approach for this PR. We can find a way to  allow _minor edits_ without deleting responses in a future PR.

Tracking _major edits_ (i.e. swap/add/delete) should be just a boolean flag on the browser side? Set the flag as soon as a swap/add/delete is done and keep it set unless the edit operation is cancelled. If the user swaps and does another swap to reverse the first one, we can still consider it as a major edit.

VamsiSangam commented:
@chowyb ready for review.

VamsiSangam commented:
Yes @damithc sir, I will create two issues after this is merged, one for improving test cases, which I will take up and another to allow simple edits to rubric choices. I plan to leave the second issue until my other PRs get merged. If nobody has claimed it, then I will work on that too (After this PR I feel like I can handle pretty much anything related to rubric questions :stuck_out_tongue_closed_eyes: )

damithc commented:
> After this PR I feel like I can handle pretty much anything related to rubric questions

@VamsiSangam It's a good strategy for you to focus on Rubric questions. Instructors love Rubric questions and our Rubric questions can be improved in many areas.

VamsiSangam commented:
Sure sir. Will keep that in mind :smile:

whipermr5 commented:
@VamsiSangam You'll need to run GodMode for the failing test due to changes after merging in #7774.

VamsiSangam commented:
@whipermr5 done!

whipermr5 commented:
@VamsiSangam Apologies; now there's one more 😅
